### PR TITLE
UI Changes for Rubric Reuse

### PR DESCRIPTION
**Pull request by jansenk**
**TL;DR -** Add a dropdown section to the rubric tab for rubric copying

JIRA: [AU-22](https://openedx.atlassian.net/browse/AU-22)

**What changed?**

- Added a dropdown to the rubric tab
- Added a "combobox"-style autocomplete to the dropdown, with all "other" ORA block ids
- Added a button that when clicked clones the specified rubric into the edit modal
- Added a message that pops up to confirm the copied rubric

**Screenshots**
base
dropdown
alert

**Developer Checklist**

- [ ] Reviewed the [release process](https://github.com/edx/edx-ora2/blob/master/.github/release_process.md)
- [ ] Translations and JS/SASS compiled
- [ ] Bumped version number in [setup.py](https://github.com/edx/edx-ora2/blob/a62e81a9b0d89223476967ec3c27f3557a850735/setup.py#L39) and [package.json](https://github.com/edx/edx-ora2/blob/a62e81a9b0d89223476967ec3c27f3557a850735/package.json#L3)

**Testing Instructions**

- Create a new course.
- Import this course: 
[course.ok28jmzh.tar.gz](https://github.com/edx/edx-ora2/files/6840710/course.ok28jmzh.tar.gz)
- In waffle flag course overrides, create a force on override for `openresponseassessment.rubric_reuse` for your course ID.
- Go to the course in studio, and navigate to the COMMAND CENTER -> USE ME TO COPY RUBRICS ora
- Click edit -> rubric
- The dropdown should be there, expanded, and you should be able to choose from a dropdown of block ids. 
- Select various blocks, and copy their  rubrics.

**Reviewer Checklist**

Collectively, these should be completed by reviewers of this PR:

- [x] I've done a visual code review
- [x] I've tested the new functionality

FYI: @edx/masters-devs-gta

### CODE REVIEW

File: openassessment/xblock/static/sass/oa/utilities/_developer.scss [hunk 1]
```diff
@@ -654,6 +654,88 @@
...
+              width: 90%;
+              height: 40px;
+            }
+
+            .openassessment_clone_combobox {
+              position: relative;
+              display: inline-block;
+            }
+
+            .openassessment_clone_combobox_toggle {
```
Comment (nsprenkle) [inline openassessment/xblock/static/sass/oa/utilities/_developer.scss:707]: Fix for weird button styling:

```suggestion
            .openassessment_clone_combobox_toggle {
              vertical-align: middle;
```

File: openassessment/templates/openassessmentblock/edit/oa_rubric_reuse.html [hunk 2]
```diff
@@ -0,0 +1,41 @@
+{% load i18n %}
+{% spaceless %}
+<div id="openassessment_rubric_clone">
+    <h3>{% trans "Clone Rubric" %}<span id="rubric-reuse-collapse-icon"></span></h3>
+    <div>
+        <script type="text/template" id="openassessment_rubric_clone_other_ora_data">{{ other_openassessments_in_course | safe }}</script>
+        <div class="openassessment_rubric_clone_block" id="openassessment_rubric_clone_description">{% trans "To clone a rubric from an existing published or unpublished draft, you can search by the Block ID. Note that cloning is one-way, meaning changes made after cloning will only effect the rubric being modified, and will not modify not any rubric it was cloned from." %}</div>
+        <div class="openassessment_rubric_clone_block" id="openassessment_rubric_clone_block_id"><b>{% trans "Block ID For this ORA:" %}</b> {{ block_location }}</div>
```
Comment (nsprenkle) [inline openassessment/templates/openassessmentblock/edit/oa_rubric_reuse.html:8]: Proposed tweak:

If you add the following CSS to the "Block ID for this ORA" you can make it non-selectable. That lets an instructor triple-click the line and highlight only the block ID.

```
user-select: none
```

**Review by nsprenkle — APPROVED** [review]
Visual & function review looks good. I proposed 2 minor CSS tweaks for better styling and usability.

Also update translations

File: openassessment/templates/openassessmentblock/edit/oa_rubric_reuse.html [hunk 3]
```diff
@@ -0,0 +1,41 @@
+{% load i18n %}
+{% spaceless %}
+<div id="openassessment_rubric_clone">
+    <h3>{% trans "Clone Rubric" %}<span id="rubric-reuse-collapse-icon"></span></h3>
+    <div>
+        <div id="openassessment_rubric_clone_description">{% trans "To clone a rubric from an existing published or unpublished draft, you can search by the Unit Name. Note that cloning is one-way, meaning changes made after cloning will only effect the rubric being modified, and will not modify not any rubric it was cloned from." %}</div>
```
[the remark this answers is not in the document]
Comment (jansenk) [reply openassessment/templates/openassessmentblock/edit/oa_rubric_reuse.html]: You're looking at old commits, sorry

File: openassessment/xblock/openassessmentblock.py [hunk 4]
```diff
@@ -541,6 +541,7 @@ def student_view(self, context=None):  # pylint: disable=unused-argument
             "rubric_assessments": ui_models,
             "show_staff_area": self.is_course_staff and not self.in_studio_preview,
         }
+        logger.info(f"------------------------------------------------------------------------------------self.is_course_staff {self.is_course_staff}")
```
[the remark this answers is not in the document]
Comment (jansenk) [reply openassessment/xblock/openassessmentblock.py]: You're looking at old commits, sorry

File: openassessment/xblock/static/js/src/studio/oa_edit_rubric.js [hunk 5]
```diff
@@ -104,6 +167,40 @@ export class EditRubricView {
...
 
+  /**
+  * Show the .openassessment_rubric_clone_alert message banner, removing one if it already exists.
+  * args:
+  *   - block_id: The block which the rubric is being cloned
+  */
+     displayRubricClonedMessage(block_id) {
+      const view = this;
+      var $rubric_clone_alerts = $('.openassessment_rubric_clone_alert', this.element);
+      if ($rubric_clone_alerts.length > 1){
```
[the remark this answers is not in the document]
Comment (jansenk) [reply openassessment/xblock/static/js/src/studio/oa_edit_rubric.js]: Is one preferable to the other? I don't know

File: openassessment/templates/openassessmentblock/edit/oa_rubric_reuse.html (hunk 2 above)
Comment (jansenk) [reply openassessment/templates/openassessmentblock/edit/oa_rubric_reuse.html:8]: That's super handy. Thanks

File: openassessment/xblock/static/sass/oa/utilities/_developer.scss (hunk 1 above)
Comment (jansenk) [reply openassessment/xblock/static/sass/oa/utilities/_developer.scss:707]: If this one little fix is all it takes .... I love css! /s

File: openassessment/xblock/static/js/src/studio/oa_edit_rubric.js [hunk 6]
```diff
@@ -104,6 +167,40 @@ export class EditRubricView {
...
+      }
+    }
+  
+    /**
+    * Helper method for displayRubricClonedMessage, constructs and displays the actual message banner
+    * args:
+    *   - block_id: The block which the rubric is being cloned
+    *   - $rubric_clone_alert_tpl: jquery object representing the hidden message template div
+    */
+    createNewRubricClonedAlert(block_id, $rubric_clone_alert_tpl){
```
[the remark this answers is not in the document]
Comment (jansenk) [reply openassessment/xblock/static/js/src/studio/oa_edit_rubric.js]: Good idea, thanks

File: openassessment/xblock/static/js/src/studio/oa_edit_rubric.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (jansenk) [reply openassessment/xblock/static/js/src/studio/oa_edit_rubric.js]: Oh you mean just don't have a template at all, and just replace the text. That makes sense. At first I had the idea that we'd stack these if you copied multiple rubrics. It doesn't really make sense when you actually think about the use case, but that's where the design came from.

File: openassessment/xblock/static/sass/oa/utilities/_developer.scss (hunk 1 above)
Comment (jansenk) [reply openassessment/xblock/static/sass/oa/utilities/_developer.scss:711]: un
Screen Shot 2021-07-23 at 1 55 37 PM
fortunately this didn't fix it

Comment (jansenk) [reply openassessment/xblock/static/sass/oa/utilities/_developer.scss:711]: It's something to do with the padding around the text input i think? I'm really not sure though

File: openassessment/xblock/studio_mixin.py [hunk 7]
```diff
@@ -183,7 +184,8 @@ def editor_context(self):
             'selected_teamset_id': self.selected_teamset_id,
             'show_rubric_during_response': self.show_rubric_during_response,
             'rubric_reuse_enabled': self.is_rubric_reuse_enabled,
-            'other_openassessments_in_course': other_openassessments_in_course,
+            'other_openassessments_in_course': json.dumps(rubric_reuse_data),
```
Comment (muselesscreator) [inline openassessment/xblock/studio_mixin.py:187]: If changing this variable name (which makes sense), why not also change key?

File: openassessment/xblock/static/js/src/studio/oa_edit_rubric.js [hunk 8]
```diff
@@ -37,6 +35,71 @@ export class EditRubricView {
...
+      source: otherOraData,
+    });
+
+    // Setup the 'combobox' dropdown button. Based on https://jqueryui.com/autocomplete/#combobox
+    let wasOpen = false;
+    $('.openassessment_clone_combobox_toggle')
+      .button()
+      .removeClass('ui-corner-all')
+      .addClass('custom-combobox-toggle ui-corner-right')
+      .on('mousedown', () => {
```
Comment (muselesscreator) [inline openassessment/xblock/static/js/src/studio/oa_edit_rubric.js:81]: so... grain of salt... but I feel like these callbacks should all just be other bound methods of the class rather than in-line paritials.

Grain of salt because this is not react code, and is not going to live for super-duper-long, but that version is easier to maintain and test, and will be easier to steal pieces of later when we attempt the final refactor.

There is a small chance that I'm missing an eccentricity about the way you are using `view=this` because there *are* specific instances where jquery kind of requires that, but we don't *seem* to be hitting any of those in this code.

File: openassessment/xblock/studio_mixin.py (hunk 7 above)
Comment (jansenk) [reply openassessment/xblock/studio_mixin.py:187]: makes sense

File: openassessment/xblock/studio_mixin.py [hunk 9]
```diff
@@ -147,9 +148,9 @@ def editor_context(self):
         white_listed_file_types_string = ','.join(white_listed_file_types) if white_listed_file_types else ''
 
         # If rubric reuse is enabled, include information about the other ORAs in this course
-        other_openassessments_in_course = []
+        rubric_reuse_data = {}
         if self.is_rubric_reuse_enabled:
-            other_openassessments_in_course = self.get_other_ora_blocks_for_rubric_editor_context()
+            rubric_reuse_data = self.get_other_ora_blocks_for_rubric_editor_context()
```
Comment (leangseu-edx) [inline openassessment/xblock/studio_mixin.py:153]: `self.get_other_ora_blocks_for_rubric_editor_context()` is still returning `list` and not `dictionary` right?

File: openassessment/xblock/static/js/src/studio/oa_edit_rubric.js [hunk 10]
```diff
@@ -37,6 +35,71 @@ export class EditRubricView {
...
+   * Sets up the UI for rubric cloning.
+   *
+   * Setup JQuery-UI Accordion dropdown panel, autocomplete input, autocomplete/combobox button.
+   * Load json representation of other course ORAs from rendered JSON
+   * Install click handler on the "Clone" button to submit request and display success banner.
+   */
+  setupRubricReuse() {
+    const view = this;
+    // If the section hasn't been rendered, rubric reuse isn't enabled, so return early
+    if ($('#openassessment_rubric_clone').size() === 0) {
```
Comment (leangseu-edx) [inline openassessment/xblock/static/js/src/studio/oa_edit_rubric.js:51]:
```suggestion
    if ($('#openassessment_rubric_clone', view.element).size() === 0) {
```

**Review by leangseu-edx — CHANGES_REQUESTED** [review]
Maybe an unpopular opinion but I think every jquery selector should have parent class eg: `this.element` as the second arguments. It prevents selecting wrong element.

File: openassessment/xblock/studio_mixin.py (hunk 9 above)
Comment (jansenk) [reply openassessment/xblock/studio_mixin.py:153]: It's a dict. I like the idea of leaving it because some day I want the way we display this to change and then we won't have to touch the backend

**Review by mattcarter — APPROVED** [review]
g2g

**Review by leangseu-edx — APPROVED** [review]
